source/patterns/@aws-solutions-constructs/aws-lambda-kinesisstreams/lib/index.ts

```diff
@@ -0,0 +1,146 @@
+    /**
+     *  Properties to override default properties if deployVpc is true
+     */
+    readonly vpcProps?: ec2.VpcProps;
+    /**
+     * Whether to deploy a new VPC
+     *
+     * @default - false
+     */
+    readonly deployVpc?: boolean;
+    /**
+     * Optional Name to override the Lambda Function default environment variable name that holds the Kinesis Data Stream name value
+     *
+     * @default - KINESIS_DATA_STREAM_NAME
+     */
+    readonly streamEnvironmentVariableName?: string;
+}
+
+/**
+ * @summary The LambdaToKinesisStream class.
+ */
+export class LambdaToKinesisStreams extends Construct {
+    public readonly vpc?: ec2.IVpc;
+    public readonly kinesisStream: kinesis.Stream;
+    public readonly lambdaFunction: lambda.Function;
+    public readonly cloudwatchAlarms?: cloudwatch.Alarm[];
+
+    /**
+     * @summary Constructs a new instance of the KinesisStreamsToLambda class.
+     * @param {cdk.App} scope - represents the scope for all the resources.
+     * @param {string} id - this is a a scope-unique id.
+     * @param {LambdaToKinesisStreamsProps} props - user provided props for the construct
+     * @since 0.8.0
+     * @access public
+     */
+    constructor(scope: Construct, id: string, props: LambdaToKinesisStreamsProps) {
+      super(scope, id);
+      defaults.CheckProps(props);
+
+      // Setup the VPC
+      if (props.deployVpc || props.existingVpc) {
+        this.vpc = defaults.buildVpc(scope, {
+          defaultVpcProps: defaults.DefaultIsolatedVpcProps(),
+          existingVpc: props.existingVpc,
+          userVpcProps: props.vpcProps,
+          constructVpcProps: {
+            enableDnsHostnames: true,
+            enableDnsSupport: true,
+          },
+        });
+
+        defaults.AddAwsServiceEndpoint(scope, this.vpc, defaults.ServiceEndpointTypes.KINESIS_STREAMS);
+      }
+
+      // Setup the Lambda function
+      this.lambdaFunction = defaults.buildLambdaFunction(this, {
+        existingLambdaObj: props.existingLambdaObj,
+        lambdaFunctionProps: props.lambdaFunctionProps,
+        vpc: this.vpc
+      });
+
+      // Setup the Kinesis Stream
+      this.kinesisStream = defaults.buildKinesisStream(this, {
+        existingStreamObj: props.existingStreamObj,
+        kinesisStreamProps: props.kinesisStreamProps
+      });
+
+      // Configure Lambda Function environment variables
+      const streamNameEnvironmentVariableName = props.streamEnvironmentVariableName || 'KINESIS_DATA_STREAM_NAME';
+      this.lambdaFunction.addEnvironment(streamNameEnvironmentVariableName, this.kinesisStream.streamName);
+
+      // Grant the Lambda Function permission to write to the Kinesis Stream
+      this.kinesisStream.grantWrite(this.lambdaFunction.grantPrincipal);
+
+      // Grant the Lambda Function permission to use the stream's encryption key so it can publish messages to it
+      this.kinesisStream.encryptionKey?.grant(this.lambdaFunction.grantPrincipal,
```

Comment:
What if the client has provided props that don't encrypt the stream, or an existing stream that is not encrypted?

Will the ? operator handle this situation? Is this the same question I had on the last construct, or just the same type of question? I might learn more as I continue to read.

Reply:
Yes this is the same pattern/question. If the `encryptionKey` property of the `kinesisStream` object is null, then the `?` operator stops execution of the line at that point, so it will not call the `grant` method. The technique is called [Optional Chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining)